src/envs/pybullet_cover.py

```diff
@@ -0,0 +1,313 @@
+"""A PyBullet version of Cover."""
+
+from typing import ClassVar, Dict, List, Sequence, Tuple
+
+import numpy as np
+import pybullet as p
+from gym.spaces import Box
+
+from predicators.src import utils
+from predicators.src.envs.cover import CoverEnv
+from predicators.src.envs.pybullet_env import PyBulletEnv, \
+    create_pybullet_block
+from predicators.src.envs.pybullet_robots import _SingleArmPyBulletRobot, \
+    create_single_arm_pybullet_robot
+from predicators.src.settings import CFG
+from predicators.src.structs import Action, Array, Object, \
+    ParameterizedOption, Pose3D, State, Type
+
+
+class PyBulletCoverEnv(PyBulletEnv, CoverEnv):
+    """PyBullet Cover domain."""
+    # Parameters that aren't important enough to need to clog up settings.py
+
+    # Option parameters.
+    _open_fingers: ClassVar[float] = 0.04
+    _closed_fingers: ClassVar[float] = 0.01
+
+    # Table parameters.
+    _table_pose: ClassVar[Pose3D] = (1.35, 0.75, 0.0)
+    _table_orientation: ClassVar[Sequence[float]] = [0., 0., 0., 1.]
+
+    # Robot parameters.
+    _ee_orn: ClassVar[Sequence[float]] = p.getQuaternionFromEuler(
+        [np.pi / 2, np.pi / 2, -np.pi])
+    _move_to_pose_tol: ClassVar[float] = 1e-7
+
+    # Object parameters.
+    _obj_len_hgt: ClassVar[float] = 0.045
+    _max_obj_width: ClassVar[float] = 0.07  # highest width normalized to this
+
+    # Dimension and workspace parameters.
+    _table_height: ClassVar[float] = 0.2
+    _y_lb: ClassVar[float] = 0.4
+    _y_ub: ClassVar[float] = 1.1
+    _robot_init_y: ClassVar[float] = (_y_lb + _y_ub) / 2
+    _pickplace_z: ClassVar[float] = _table_height + _obj_len_hgt * 0.5 + 0.01
+    _target_height: ClassVar[float] = 0.0001
+
+    def __init__(self) -> None:
+        super().__init__()
+
+        # Override PickPlace option
+        types = self._PickPlace.types
+        params_space = self._PickPlace.params_space
+        toggle_fingers_func = lambda s, _1, _2: (
+            (self._open_fingers, self._closed_fingers)
```

Comment:
this feels a bit wrong to me because the first argument here is not going to be the real "current" finger state, which is what is expected by the option creator. if you ever see behavior where the robot hits the object as it's trying to grasp it, leading to undesired rotations etc, then this could be to blame

Reply:
okay I think I get why this is happening though -- it's because the finger joint state isn't actually in the State, right?

Reply:
right yeah. i agree with your statement, but i never saw any behavior like that. i'm inclined to leave this as is since it seems to work fine, but i'll be on the lookout if we see this behavior in the future. i'll put a comment.

also, i'll add a video to this PR